### ISSUE TITLE
mouse input lag under X11 

### ISSUE THREAD

**Issue by benoit-pierre** [adventuregamestudio/ags#162]
When using a high polling rate mouse on Linux with X11, the mouse cursor lags. After investigation, it turns out this is due to the fact that Allegro will only process up-to 5 X11 events at a time, and so mouse motion events will pile up and the mouse cursor will lag (as reported [here](http://sourceforge.net/p/alleg/bugs/396/)).

It's also possible to fix this in AGS by setting a custom X11 input handler as seen here: df953ff183ee5ecc456880c5ba2c5659dcd26641.

**Comment by ivan-mogilko** [comment]
I have a question: if Allegro authors had deliberately limited number of events to parse per turn, perhaps they had a reason to do so. Cannot this patch make things worse in some case? Maybe there should be some condition to turn it on?

**Comment by shmerl** [comment]
Sorry for an off-topic, but do you have a plan to switch to SDL from Allegro? It would be much better in the long term for these kind of issue. Consider also Wayland support.

**Comment by benoit-pierre** [comment]
@ivan-mogilko : unfortunately looking at the VCS history does not give any clue as to why this was done like this, but event processing is not limited either in the more recent version (e.g. 5.1).

**Comment by ivan-mogilko** [comment]
> Sorry for the off-topic, but do you have a plan to switch to SDL from Allegro? It would be much better in the long term for these kind of issue. Consider also Wayland support.

This was in consideration since we started to work on this project; we were thinking about SDL, and Allegro 5, but I could never find time to actually do that, because there I had a lot of requests about more "high-level" engine features all the time.

> but event processing is not limited either in the more recent version (e.g. 5.1).
Okay, it is good to know that.

**Comment by rofl0r** [comment]
> At the moment I am not even sure how much more I will continue working on this, and if I will at all.
@ivan-mogilko I very much appreciate your excellent lead on this project, and your thoughtful decisions.
It would be a pity to see you quit.
I hope you will continue your work here for the benefit of the community, and I'd like to thank you for what you achieved so far.

**Comment by ivan-mogilko** [comment]
Implemented this patch as 6a35ec9936b545cbb4d405309f24359fbe6e80f5 (code is generally same, just put into different place).

**Comment by ivan-mogilko** [comment]
It was found that adding the patch to AGS (original: https://github.com/adventuregamestudio/ags/commit/df953ff183ee5ecc456880c5ba2c5659dcd26641, implemented as: 6a35ec9936b545cbb4d405309f24359fbe6e80f5) actually causes tremendous delays between the game is redrawn, if the game *is run in a window size equal to its native resolution* (without scaling filter).

Debugging shows that in such case the allegro's input handling routine is being called 20 or more times in a row, hinting that the event queue is very large (or that program logic is broken somewhere).

I tried applying same patch over the master branch to see if this has something to do with latest months changes, but the result was the same.

Since apparently this patch was being used for quite a while already, I would like to clarify, did you use the patch to AGS, or patch to Allegro, and did any of you ever tried to run the game without scaling filter while having this patch on?

**Comment by benoit-pierre** [comment]
Is that only when moving the mouse cursor? Are you using a high polling rate mouse? If I remember correctly (it has been a while...), I was using a patch with a pre-loaded shared library with a Steam version of an AGS game, and also did play a number of other games with a patched version of Allegro and a custom version of AGS (compiled locally to play some Windows only games). I'm pretty sure I was playing in windowed mode, but I don't remember if I was using a scaling filter or not...

**Comment by benoit-pierre** [comment]
What would be the content of  `acsetup.cfg` for your use case?

**Comment by ivan-mogilko** [comment]
It just happens on its own, I may not touch the mouse at all.

I use a game with animated introduction for a test, and the animation progresses, but game frame is got redrawn only once in 5-8 seconds.
It occurs in both windowed and fullscreen modes, but only if no scaling filter is on. That seem to be the only condition, rest of configuration does not seem to be important.

If I comment out a line which assigns a custom XWin input handler, everything gets back to normal.

**Comment by benoit-pierre** [comment]
Which game is that? I just tried some games with `ags --windowed --gfxfilter None` and had no issue (that's with my version of AGS/Allegro). Did you check what type of events you were getting?

**Comment by ivan-mogilko** [comment]
I may try on several other games shortly, but I was not the only person who had this issue.

What I wanted to ask, are you sure it is correct to rely on XQLength when repeatedly calling "_xwin_private_input_handler"? May there be types of events in that queue that allegro's input handler does not handle (therefore calling it again may be redundant)?

I will try to print some info from the event queue, but I need to read docs on XWindows first :-).

**Comment by ivan-mogilko** [comment]
Update: yes I can confirm that this happens with other games too.

**Comment by benoit-pierre** [comment]
I would try with a patched Allegro first, as this is were the fix really belongs.

**Comment by ivan-mogilko** [comment]
I guess I may have made a mistake by putting this patch into AGS itself.

**Comment by benoit-pierre** [comment]
OK, so after updating AGS to 3.3.5.2, I can reproduce the problem. Without the `_xwin_private_input_handler` patch in AGS, and with the [patch](http://sourceforge.net/p/alleg/bugs/396/) in Allegro, I don't get any slowdown.

**Comment by benoit-pierre** [comment]
And the mouse is working fine too, without the patch in Allegro, the cursor lag is back.

**Comment by stqn1** [comment]
It looks like this is another bug that would be fixed by switching from Allegro to SDL2.
(Sorry for the pro-SDL2 spam, I’ll stop here for today ;).)

**Comment by ivan-mogilko** [comment]
There was no need for "pro-SDL2" spam anyway. It's been in plans for a while, but moving there requires quite a bit of work. @sonneveld  is trying to make this change currently in their personal fork, in steps, but that may take a while to get everything set up.

**Comment by rofl0r** [comment]
tbh i would prefer SDL(1) over SDL2 a lot since 
1) it is already mature 
2) there's a ton of games using it and it not likely that they will ever update 
3) so you will need to have SDL installed anyway 
4) so it would be kinda wasteful to have to install SDL2 just for a single game/engine when all the other stuff you use is based on SDL1 
5) once you have the SDL1 code adding optional support for SDL2 is probably pretty easy since they're pretty similar and just needs some small changes under `#ifdef`.

so it is basically the same situation with GTK+3 vs GTK+2 and python3 vs python2.7.

**Comment by shmerl** [comment]
SDL1 is not supported upstream and is already deprecated. It never should be used for anything new for sure. It doesn't work with Wayland and is basically EOL already.

**Comment by rofl0r** [comment]
> I guess the mouse just generates more events.

that sounds plausible. maybe the mouse emits so many events that the event buffer (maximum amount of events that can be saved between polls) overflows and so the oldest events are being discarded. in that case calling the input event poll function more often might fix the issue, as well as increasing the size of the event buffer (if possible).

**Comment by morganwillcock** [comment]
I don't think that events are being discarded, just there are just too many to process in real-time. The path for cursor motion is 'correct', but does not keep up with the physical motion of the mouse. I guess if it is possible to track the rate of events and differentiate between movement and a click, you could only accept a limited number of movement events (per tick?) into the event queue, but keep all click events.

**Comment by 256colour** [comment]
I've read through this thread a couple of times, and I don't quite follow if/how this problem was addressed in the end? We had a couple of testers report the same sluggish mousing problem on their high-polling-rate mice, and now need to address the bug with a patch.

I've applied the Allegro mouse patch [referenced here](
http://sourceforge.net/p/alleg/bugs/396/) and manually compiled Allegro 4.4.2. I am hesitant to release the patched Allegro to our users if there is a known problem with the fix.

Furthermore, is there a process for working/including Allegro patches into the AGS source distribution? I rely upon the /ags/debian/make_ags+libraries.sh script, and it does not (from what I can see) make a provision for manually patching Allegro before it is compiled.

**Comment by morganwillcock** [comment]
As far as I know it is still an issue, depending on the type of device that is being used. It'll work fine if a mouse with a high polling rate is connected, but not if that is the mouse you are generating input with (i.e. a second mouse on the same system might have no issues).

Asesprite on Linux is building Allegro with X_MAX_EVENTS defined as 50 instead of 5, and with cursor speed functions removed (this might be unrelated), but I think the general consensus was that getting off Allegro 4 is the only way to have it work for everyone. I don't know what state the SDL2 port is in, or even if it is actively worked on, but it might be worth testing the game against as a potential solution.

> I've applied the Allegro mouse patch referenced here and manually compiled Allegro 4.4.2. I am hesitant to release the patched Allegro to our users if there is a known problem with the fix.

The change that was previously made was meant to be functionally equivalent to this patch, but seemed to have negative effect for most people who tried it. It was reverted [here](https://github.com/adventuregamestudio/ags/pull/317/files).

>Furthermore, is there a process for working/including Allegro patches into the AGS source distribution?

I believe the fork of Allegro was originally for Windows specific changes, and other ports dealt with the patches separately, so there would be no mechanism for patching per platform. The Linux build script just expects that you already have all of the pre-requisites.

**Comment by rofl0r** [comment]
we are currently waiting on the guys from allegro to make a release ( https://github.com/liballeg/allegro5/pull/972 ) since apparently all improvements from ags have been upstreamed, so there's no reason anymore to build a special allegro for ags (even though i personally used vanilla allegro in the past and it worked reasonably well).

> Asesprite on Linux is building Allegro with X_MAX_EVENTS defined as 50 instead of 5

it may be a good idea to ask @SiegeLord to add this change to allegro before the release, unless it was already upstreamed too.

@256colour  is there a specific reason that you use a special fork instead of the version here ?

**Comment by SiegeLord** [comment]
Sure, please send a PR for anything you want upstreamed. It doesn't appear that `X_MAX_EVENTS` change has been.

**Comment by rofl0r** [comment]
> Sure, please send a PR for anything you want upstreamed. It doesn't appear that X_MAX_EVENTS change has been.

thanks!
https://github.com/liballeg/allegro5/pull/984

**Comment by 256colour** [comment]
@rofl0r Ah, no - I just applied that patch because it was the first one I came across to deal with the problem.

**Comment by shmerl** [comment]
This affects Mage's Initiation. The mouse is very laggy (GOG release).

**Comment by shmerl** [comment]
@256colour: are you the developer of Mage's Initiation? Can you please update the GOG release with the Allegro's mouse lag fix?

**Comment by ivan-mogilko** [comment]
We have a SDL2-based version now, this problem has to be retested in it.

**Comment by ericoporto** [comment]
> Playing a very high resolution game (so no scaling is in effect) with a mouse that has a switchable tracking speed seems to result in very sluggish and delayed movement

@morganwillcock do you still have that mouse?

**Comment by morganwillcock** [comment]
> do you still have that mouse?

I do! If you point me in the direction of what to test I can check something.

**Comment by ericoporto** [comment]
@morganwillcock  I don't have the issue unfortunately, but what I was thinking is if you could reproduce the issue you commented years ago here: https://github.com/adventuregamestudio/ags/issues/162#issuecomment-389978756

**Comment by morganwillcock** [comment]
I tried the master branch, building it with CMake, and I couldn't get any lag with any device. So presumably on SDL2 this is not an issue.

**Comment by ericoporto** [comment]
Thanks! Did the lag appeared when using the non-SDL2 version of the engine?

**Comment by morganwillcock** [comment]
I cannot, but the Allegro4 version that comes with Debian is now 4.4.3 and this has the same changes to the event limits that the version in the AGS repository has.

**Comment by ivan-mogilko** [comment]
I think it's safe to close this then. In any case if something is wrong in 3.6.0 it should be opened as a new issue, related to SDL2.
